File: src/main/java/org/micoli/php/service/filesystem/FileListProcessor.java
package org.micoli.php.service.filesystem;

import com.intellij.openapi.vfs.VfsUtil;
import com.intellij.openapi.vfs.VirtualFile;
import java.io.*;
import java.util.*;
import org.eclipse.jgit.ignore.IgnoreNode;
import org.jetbrains.annotations.Nullable;

public class FileListProcessor {

    public static List<VirtualFile> findFilesFromSelectedFiles(List<VirtualFile> selectedFiles) {
        Set<VirtualFile> filesSet = new LinkedHashSet<>();

        for (VirtualFile file : selectedFiles) {
            if (file == null) {
                continue;
            }
            if (file.isDirectory()) {
                listFilesRecursively(file, filesSet);
                continue;
            }
            filesSet.add(file);
        }
        return new ArrayList<>(filesSet);
    }

    public static List<VirtualFile> filterFiles(@Nullable VirtualFile ignoreFile, List<VirtualFile> filesSet) {
        IgnoreNode ignoreNode = getIgnoreNode(ignoreFile);
        if (ignoreNode == null) {
            return new ArrayList<>(filesSet);
        }

        return filesSet.stream()
                .filter(file -> ignoreNode.isIgnored(VfsUtil.getRelativePath(file, ignoreFile.getParent()), false)
                        != IgnoreNode.MatchResult.IGNORED)
                .toList();
    }

    private static @Nullable IgnoreNode getIgnoreNode(VirtualFile ignoreFile) {
        if (ignoreFile == null) {
            return null;
        }
        try {
            final IgnoreNode ignoreNode = new IgnoreNode();
            try (InputStream reader = ignoreFile.getInputStream()) {
                ignoreNode.parse(reader);
            }
            return ignoreNode;
        } catch (IOException ignored) {
            return null;
        }
    }

    private static void listFilesRecursively(VirtualFile directory, Set<VirtualFile> filesList) {
        VirtualFile[] children = directory.getChildren();

        if (children == null) {
            return;
        }
        for (VirtualFile child : children) {
            if (child.isDirectory()) {
                listFilesRecursively(child, filesList);
                continue;
            }
            filesList.add(child);
        }
    }
}

File: src/main/kotlin/org/micoli/php/service/filesystem/FileListProcessor.kt
package org.micoli.php.service.filesystem

import com.intellij.openapi.vfs.VfsUtil
import com.intellij.openapi.vfs.VirtualFile
import java.io.IOException
import org.eclipse.jgit.ignore.IgnoreNode

object FileListProcessor {
    @JvmStatic
    fun findFilesFromSelectedFiles(
        selectedFiles: MutableList<VirtualFile>
    ): MutableList<VirtualFile> {
        val filesSet: MutableSet<VirtualFile> = LinkedHashSet()

        for (file in selectedFiles) {
            if (file.isDirectory) {
                listFilesRecursively(file, filesSet)
                continue
            }
            filesSet.add(file)
        }
        return ArrayList(filesSet)
    }

    @JvmStatic
    fun filterFiles(
        ignoreFile: VirtualFile?,
        filesSet: MutableList<VirtualFile>
    ): MutableList<VirtualFile> {
        val ignoreNode = getIgnoreNode(ignoreFile) ?: return ArrayList(filesSet)

        return filesSet
            .stream()
            .filter { it != null }
            .filter {
                (ignoreNode.isIgnored(VfsUtil.getRelativePath(it, ignoreFile!!.parent), false) !=
                    IgnoreNode.MatchResult.IGNORED)
            }
            .toList()
    }

    private fun getIgnoreNode(ignoreFile: VirtualFile?): IgnoreNode? {
        if (ignoreFile == null) {
            return null
        }
        try {
            val ignoreNode = IgnoreNode()
            ignoreFile.inputStream.use { reader -> ignoreNode.parse(reader) }
            return ignoreNode
        } catch (_: IOException) {
            return null
        }
    }

    private fun listFilesRecursively(directory: VirtualFile, filesList: MutableSet<VirtualFile>) {
        val children = directory.children ?: return

        for (child in children) {
            if (child.isDirectory) {
                listFilesRecursively(child, filesList)
                continue
            }
            filesList.add(child)
        }
    }
}
